PyPy: Rect test failures

With pypy:

```
======================================================================
ERROR: test_clamp (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 750, in test_clamp
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
ERROR: test_clip (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 757, in test_clip
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
ERROR: test_copy (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 729, in test_copy
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
ERROR: test_fit (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 778, in test_fit
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
ERROR: test_inflate (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 743, in test_inflate
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
ERROR: test_move (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 736, in test_move
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
ERROR: test_union (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 764, in test_union
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
ERROR: test_unionall (test.rect_test.SubclassTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 771, in test_unionall
    self.assertTrue(mr1.an_attribute)
AttributeError: 'pygame.Rect' object has no attribute 'an_attribute'

======================================================================
FAIL: testEquals (test.rect_test.RectTypeTest)
check to see how the rect uses __eq__
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Documents\dev\pygame\test\rect_test.py", line 513, in testEquals
    self.assertEqual(r6, r2)
AssertionError: <rect(0, 0, 0, 0)> != <rect(10, 20, 30, 40)>
```

**Related Docs:** https://www.pygame.org/docs/ref/rect.html

--- a/src_c/rect.c
+++ b/src_c/rect.c
@@ -35,7 +35,8 @@
 #include <limits.h>
 
 static PyTypeObject pgRect_Type;
-#define pgRect_Check(x) ((x)->ob_type == &pgRect_Type)
+#define pgRect_Check(x) (PyObject_IsInstance(x, (PyObject *)&pgRect_Type))
+#define pgRect_CheckExact(x) (Py_TYPE(x) == &pgRect_Type)
 
 static int
 pg_rect_init(pgRectObject *, PyObject *, PyObject *);
@@ -174,7 +175,7 @@ pg_tuple_from_values_int(int val1, int val2)
 static PyObject *
 _pg_rect_subtype_new4(PyTypeObject *type, int x, int y, int w, int h)
 {
-    pgRectObject *rect = (pgRectObject *)pgRect_Type.tp_new(type, NULL, NULL);
+    pgRectObject *rect = type->tp_new(type, NULL, NULL);
 
     if (rect) {
         rect->r.x = x;
@@ -191,7 +192,9 @@ pg_rect_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
     pgRectObject *self;
 
 #ifdef PYPY_VERSION
-    if (pg_rect_freelist_num > -1) {
+    /* Only instances of the base pygame.Rect class are allowed in the
+     * current freelist implementation (subclasses are not allowed) */
+    if (pg_rect_freelist_num > -1 && type == &pgRect_Type) {
         self = pg_rect_freelist[pg_rect_freelist_num];
         Py_INCREF(self);
         /* This is so that pypy garbage collector thinks it is a new obj
@@ -225,7 +228,10 @@ pg_rect_dealloc(pgRectObject *self)
     }
 
 #ifdef PYPY_VERSION
-    if (pg_rect_freelist_num < PG_RECT_FREELIST_MAX - 1) {
+    /* Only instances of the base pygame.Rect class are allowed in the
+     * current freelist implementation (subclasses are not allowed) */
+    if (pg_rect_freelist_num < PG_RECT_FREELIST_MAX - 1 &&
+        pgRect_CheckExact(self)) {
         pg_rect_freelist_num++;
         pg_rect_freelist[pg_rect_freelist_num] = self;
     }
@@ -1463,7 +1469,7 @@ pg_rect_ass_subscript(pgRectObject *self, PyObject *op, PyObject *value)
             self->r.w = val;
             self->r.h = val;
         }
-        else if (PyObject_IsInstance(value, (PyObject *)&pgRect_Type)) {
+        else if (pgRect_Check(value)) {
             pgRectObject *rect = (pgRectObject *)value;
 
             self->r.x = rect->r.x;

--- a/src_c/rect.c
+++ b/src_c/rect.c
@@ -175,7 +175,7 @@ pg_tuple_from_values_int(int val1, int val2)
 static PyObject *
 _pg_rect_subtype_new4(PyTypeObject *type, int x, int y, int w, int h)
 {
-    pgRectObject *rect = type->tp_new(type, NULL, NULL);
+    pgRectObject *rect = (pgRectObject *)type->tp_new(type, NULL, NULL);
 
     if (rect) {
         rect->r.x = x;

--- a/test/rect_test.py
+++ b/test/rect_test.py
@@ -1,12 +1,9 @@
 import math
-import platform
 import unittest
 from collections.abc import Collection, Sequence
 
 from pygame import Rect, Vector2
 from pygame.tests import test_utils
-
-IS_PYPY = "PyPy" == platform.python_implementation()
 
 
 class RectTypeTest(unittest.TestCase):
@@ -69,7 +66,6 @@ class RectTypeTest(unittest.TestCase):
 
         self.assertEqual(test_rect, expected_normalized_rect)
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy sometimes")
     def test_normalize__positive_height(self):
         """Ensures normalize works with a negative width and a positive height."""
         test_rect = Rect((1, 2), (-3, 6))
@@ -82,7 +78,6 @@ class RectTypeTest(unittest.TestCase):
 
         self.assertEqual(test_rect, expected_normalized_rect)
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy sometimes")
     def test_normalize__positive_width(self):
         """Ensures normalize works with a positive width and a negative height."""
         test_rect = Rect((1, 2), (3, -6))
@@ -95,7 +90,6 @@ class RectTypeTest(unittest.TestCase):
 
         self.assertEqual(test_rect, expected_normalized_rect)
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy sometimes")
     def test_normalize__zero_height(self):
         """Ensures normalize works with a negative width and a zero height."""
         test_rect = Rect((1, 2), (-3, 0))
@@ -108,7 +102,6 @@ class RectTypeTest(unittest.TestCase):
 
         self.assertEqual(test_rect, expected_normalized_rect)
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy sometimes")
     def test_normalize__zero_width(self):
         """Ensures normalize works with a zero width and a negative height."""
         test_rect = Rect((1, 2), (0, -6))
@@ -121,7 +114,6 @@ class RectTypeTest(unittest.TestCase):
 
         self.assertEqual(test_rect, expected_normalized_rect)
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy")
     def test_normalize__non_negative(self):
         """Ensures normalize works when width and height are both non-negative.
 
@@ -1264,7 +1256,6 @@ class RectTypeTest(unittest.TestCase):
             with self.assertRaises(TypeError):
                 clipped_line = rect.clipline(*line)
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy sometimes")
     def test_move(self):
         r = Rect(1, 2, 3, 4)
         move_x = 10
@@ -1273,7 +1264,6 @@ class RectTypeTest(unittest.TestCase):
         expected_r2 = Rect(r.left + move_x, r.top + move_y, r.width, r.height)
         self.assertEqual(expected_r2, r2)
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy sometimes")
     def test_move_ip(self):
         r = Rect(1, 2, 3, 4)
         r2 = Rect(r)
@@ -1404,7 +1394,6 @@ class RectTypeTest(unittest.TestCase):
             "r1 collides with Rect(r1.right, r1.bottom, 1, 1)",
         )
 
-    @unittest.skipIf(IS_PYPY, "fails on pypy3 sometimes")
     def testEquals(self):
         """check to see how the rect uses __eq__"""
         r1 = Rect(1, 2, 3, 4)
@@ -2546,7 +2535,6 @@ class RectTypeTest(unittest.TestCase):
         self.assertFalse(isinstance(r, Sequence))
 
 
-@unittest.skipIf(IS_PYPY, "fails on pypy")
 class SubclassTest(unittest.TestCase):
     class MyRect(Rect):
         def __init__(self, *args, **kwds):

